### PR TITLE
Improved buffering

### PR DESCRIPTION
**Pull request by akhenry**
NOTE: https://github.com/nasa/openmct/pull/7837 will need be merged before this.
1. Merge https://github.com/nasa/openmct/pull/7837
2. Update [package.json](https://github.com/akhenry/openmct-yamcs/pull/477/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R22) to point it to npm `latest`

Adapts the openmct-yamcs plugin for the API changes in https://github.com/nasa/openmct/issues/7838

### Describe your changes:
* Modifies telemetry subscription handling in openmct-yamcs adapter to support improvements to the telemetry API in Open MCT.
* [Removes the Yamcs-specific batching strategy previously required ](https://github.com/nasa/openmct/pull/7837/files#diff-ad7bb0f2f80ab9afc9c3eadbe47859ed40094b6a487141fd2a72bffeb216e62cL145). Open MCT now uses a more flexible one-size fits all strategy
* [Adds a new test to catch the failure mode identified in the previous strategy]().

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this a [notable change](../docs/src/process/release.md) that will require a special callout in the release notes? For example, will this break compatibility with existing APIs or projects that consume these plugins?

### Author Checklist

* [x] Changes address original issue?
* [x] Tests included and/or updated with changes?
* [x] Has this been smoke tested?
* [x] Have you associated this PR with a `type:` label? Note: this is not necessarily the same as the original issue.
* [ ] Have you associated a milestone with this PR? Note: leave blank if unsure.
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Reviewer has tested changes by following the provided instructions?
* [x] Changes appear not to be breaking changes?
* [x] Appropriate automated tests included?
* [x] Code style and in-line documentation are appropriate?

### CODE REVIEW

File: tests/e2e/yamcs/realtimeData.e2e.spec.mjs [hunk 1]
```diff
@@ -249,6 +251,101 @@ test.describe('Realtime telemetry displays', () => {
             }
         });
 
+        test.only('Open MCT does not drop telemetry when a burst of telemetry arrives', async ({ page }) => {
```
Comment (unlikelyzero) [inline tests/e2e/yamcs/realtimeData.e2e.spec.mjs:254]: test('Does not drop telemetry when a burst of telemetry greater than the buffer arrives')

File: tests/e2e/yamcs/realtimeData.e2e.spec.mjs [hunk 2]
```diff
@@ -249,6 +251,101 @@ test.describe('Realtime telemetry displays', () => {
...
+            const PARAMETER_VALUES_COUNT = 60;
+            /**
+             * A failure mode of the previous implementation of batching was when bursts of telemetry from a parameter arrived all at once.
+             * A burst of 60 messages will overwhelm a per-parameter telemetry buffer of 50, but will not overwhelm a larger shared buffer.
+             */
+
+            // Disable real playback. We are going to inject our own batch of messages
+            await disableLink(yamcsURL);
+
+            const callNumber = await page.evaluate(async () => {
```
Comment (unlikelyzero) [inline tests/e2e/yamcs/realtimeData.e2e.spec.mjs:264]: Can you update this variable name and doc?

File: tests/e2e/yamcs/realtimeData.e2e.spec.mjs [hunk 3]
```diff
@@ -307,7 +407,7 @@ test.describe('Realtime telemetry displays', () => {
```
Comment (unlikelyzero) [inline tests/e2e/yamcs/realtimeData.e2e.spec.mjs:407]: would it be beneficial to add the performance plugin and read the value?

File: tests/e2e/yamcs/realtimeData.e2e.spec.mjs [hunk 4]
```diff
@@ -249,6 +251,101 @@ test.describe('Realtime telemetry displays', () => {
...
+            });
+
+            expect(telemetryValues.length).toBe(PARAMETER_VALUES_COUNT);
+
+            const notification = page.getByRole('alert');
+            const count = await notification.count();
+
+            if (count > 0) {
+                const text = await notification.innerText();
+                expect(text).not.toBe('Telemetry dropped due to client rate limiting.');
```
Comment (unlikelyzero) [inline tests/e2e/yamcs/realtimeData.e2e.spec.mjs:343]: Would it be beneficial to have added the performanceIndicator plugin and read the value here?

File: tests/e2e/yamcs/realtimeData.e2e.spec.mjs (hunk 1 above)
Comment (akhenry) [reply tests/e2e/yamcs/realtimeData.e2e.spec.mjs:254]: Good call, will do.

File: tests/e2e/yamcs/realtimeData.e2e.spec.mjs (hunk 2 above)
Comment (akhenry) [reply tests/e2e/yamcs/realtimeData.e2e.spec.mjs:264]: Yep, both could be better here, will do.

File: tests/e2e/yamcs/realtimeData.e2e.spec.mjs (hunk 4 above)
Comment (akhenry) [reply tests/e2e/yamcs/realtimeData.e2e.spec.mjs:363]: The original user report calls out this notification specifically, so I think it's a reasonable test to validate we have fixed the issue as reported.

[I have added a test to the open source repository to test the new performance overlay though.](https://github.com/nasa/openmct/pull/7837/files#diff-71cc5c9ef53eb1f19f8970f4cac9dbd45efaa6db2e2d5ffa3715936e064c39eaR46)

**Review by shefalijoshi — CHANGES_REQUESTED** [review]
This looks fine overall, however, the staleness and remote clock changes should be moved out into separate PRs.

**Comment by akhenry** [comment]
Test was failing due to genuine bug introduced in this refactor of telemetry batching. Previously housekeeping WebSocket messages bypassed the throttling / batching process altogether, which meant that important housekeeping messages (such as those that go back and forth when setting up a telemetry subscription) were never dropped.

With this refactor the WebSocket buffering code was greatly simplified, removing the special handling for housekeeping messages. One of the automated tests triggered an edge case where throttling is induced during the initial handshaking that takes place when initializing a subscription, causing some subscriptions to never get established properly. Luckily our tests caught this.

I made a small change to the telemetry throttling implementation so that instead of throttling all messages when the UI thread is overloaded, a regex can optionally be supplied that is used to decided whether a given message should be throttled. This allows us to avoid culling housekeeping messages.

**Review by shefalijoshi — APPROVED** [review]
Looks good
